Add cache control headers for OCW

ibrahimjaved12 opened the pull request:
# What are the relevant tickets?
Partial solution to https://github.com/mitodl/ocw-hugo-themes/issues/1212

# Description (What does it do?)
Adds cache-control headers for OCW resources.

### The pattern:
`(bereq.url.path ~ ".*(main|common|www|course_v2|instructor_insights|fields)\.[0-9a-f]+\.(css|js)\z") || (bereq.url.path ~ ".*[0-9a-f]+\.[0-9a-f]+\.(css|js)\z")`
The first part of the pattern corresponds to hashed themes related `js` or `css` files.
The second part of the pattern corresponds to dynamic imports using webpack (videojs and nanogallery2).
Eg. 
`https://ocw.mit.edu/static_shared/js/common.904c3.js (1st part)
`
`https://ocw.mit.edu/static_shared/js/abc354.904c3.js (1st part)
`
`https://ocw.mit.edu/static_shared/js/354.904c3.js (2nd part)
`
You can notice the dynamic imports here:
https://ocw.mit.edu/courses/16-885j-aircraft-systems-engineering-fall-2005/pages/related-resources/
<img width="1263" alt="image" src="https://github.com/mitodl/ol-infrastructure/assets/109785089/70c62265-c6bc-4bd7-a8e3-8e09b64aa580">
The code for them is given here:
https://github.com/mitodl/ocw-hugo-themes/blob/main/course-v2/assets/course-v2.ts

### The pattern
`(bereq.url.path ~ ".*/static_shared/images/.*\.[0-9a-f]+\.(png|jpg|jpeg|svg|gif)\z") || (bereq.url.path ~ ".*/static_shared/fonts/.*\.subset\.[0-9a-f]+\.(ttf|woff|woff2)\z")`
Corresponds to hashed static assets (images and fonts).
Eg.
`http://localhost:3000/static_shared/images/ocw_logo_white_v2.1a5cf4568192c88b5efab45063b86d8a.svg
`
`http://localhost:3000/static_shared/fonts/MaterialIcons-Regular.subset.0ec6028d1bb7a2675394e0512f8ceee4.woff2 
`

### The pattern
`(bereq.url.path ~ ".*/static_shared/mathjax/.*\.[0-9a-f]+.*\z")`
Corresponds to static mathjax files. There are many different files with different extensions produced by webpack, so we are not limiting the extensions here in the pattern.

### The pattern
`(bereq.url.path ~ ".*/static_shared/(images|fonts)/.*\.(png|jpg|jpeg|svg|gif|ttf|woff|woff2)\z")`
Corresponds to unhashed static images and fonts. Currently I do not think we have any such case, as we have hashed all the static images and fonts by webpack. But if we did, it would fall here.

### The pattern
`(bereq.url.path ~ ".*/(ocw-www|courses|images)/.*\.(png|jpg|jpeg|svg|gif)\z")`
Corresponds to any other images present in `ocw-www` or `course-v2` theme.
Eg. 
https://ocw.mit.edu/images/homepage_hero.jpg
https://ocw.mit.edu/courses/6-033-computer-system-engineering-spring-2018/a79fa5
https://ocw.mit.edu/ocw-www/chansa_20230113.jpg

### The pattern
`(bereq.http.Host ~ "www.ocw-openmatters.org") && (bereq.url.path ~ "^/wp-content/uploads/.*\.(png|jpg|jpeg|svg|gif)\z")`
`

Corresponds to any OpenMatters images we use. We use them in `ocw-www` theme, eg. this image:
https://www.ocw-openmatters.org/wp-content/uploads/2020/01/Placeholder_2.png

# How can this be tested?
I will be testing this in RC as I do not have the setup of `ol-infrastructure`.
What you need to do is, make sure the code works as intended (take the given examples for help). 

# Additional Context
More details can be found here:
https://github.com/mitodl/ocw-hugo-themes/issues/1212
https://github.com/mitodl/ocw-hugo-themes/pull/1272
https://mitodl.slack.com/archives/GQ89D0Z5M/p1696431334601819

<!--- Uncomment and add steps to be completed before merging this PR if necessary
## Checklist:
- [ ] e.g. Update secret values in Vault before merging
--->

pdpinch commented on src/ol_infrastructure/applications/ocw_site/snippets/ttl_setup.vcl:
```diff
@@ -2,9 +2,17 @@ if (beresp.status == 404) {
+} elsif (bereq.url.path ~ ".*\/static_shared\/(images|fonts)\/.*\.(png|jpg|jpeg|svg|ttf|woff|woff2)") {
+  set beresp.ttl = 604800s;  // one week
+  set beresp.http.Cache-Control = "max-age=604800";
+}
```
Why are some assets one month and others are one week?

I'm not just asking why, but I'm also trying to anticipate the next engineer or devops who looks at this and wonders "why is this?" -- how will they find out?

ibrahimjaved12 replied:
>Why are some assets one month and others are one week?
I'm not just asking why, but I'm also trying to anticipate the next engineer or devops who looks at this and wonders "why is this?" -- how will they find out?

The hashed static assets are set to one month, because they will always have fresh content.
The unhashed static assets are set to one week, because they may contain stale content, and the fresh copy will only be fetched after a week is over. There are only about 2 images that fall into this category (that could not be hashed because they were in `.ts` files). So a generalized rule is made which would be applicable to any future unhashed assets as well.

More details can be found here:
https://github.com/mitodl/ocw-hugo-themes/issues/1212#issuecomment-1743506321
https://mitodl.slack.com/archives/GQ89D0Z5M/p1696431334601819

pdpinch commented on src/ol_infrastructure/applications/ocw_site/snippets/ttl_setup.vcl:
```diff
@@ -2,9 +2,17 @@ if (beresp.status == 404) {
+}
 
-if (bereq.url.path ~ "^/main\.[0-9a-f]+\.(css|js)") {
+if (bereq.url.path ~ ".*(main|common|www|course_v2|instructor_insights|fields|[0-9a-f]+)\.[0-9a-f]+\.(css|js)") {
```
Is there supposed to be a | "or" between `fields` and the hash match, `[0-9a-f]+)\.` ?

Why do we have the hash match twice? 

I know this is out of scope, but is there any reason for the parentheses in `(css|js)` -- aren't square brackets sufficient?

pdpinch commented:
General comment: I'm wondering if we might make this clearer by using multiple conditionals instead of one regex with conditionals in it. 

I'm not sure this is a good idea. It's just something worth considering.

ibrahimjaved12 commented:
This is currently blocked by https://github.com/mitodl/ocw-studio/pull/2017 & https://github.com/mitodl/ocw-hugo-themes/pull/1272

ibrahimjaved12 commented:
Sorry for the delay; I've updated the PR instructions now and also added another case for `mathjax`.
This PR is ready to be reviewed now - the `pre-commit.ci` check is failing because of some other issue.

feoh commented:
@ibrahimjaved12 Please rebase from main and your pre-commit issue should be fixed.

pdpinch commented on src/ol_infrastructure/applications/ocw_site/snippets/ttl_setup.vcl:
```diff
@@ -2,9 +2,30 @@ if (beresp.status == 404) {
-if (bereq.url.path ~ "^/main\.[0-9a-f]+\.(css|js)") {
-  set beresp.ttl = 2629743s;  // one month
+}
+# One month cache is for hashed static assets (first three cases), while one week is for unhashed assets (last three cases)
```
Can you be more explicit about the rationale and impact of these decisions? Perhaps something like: 

```suggestion
# One month cache is for hashed static assets (first three cases). We use a long cache because the hashes should be updated whenever the files are updated. 
# One week cache is for unhashed assets (last three cases). We use a shorter cache to be reasonably confident this files are fresh
```

Why one month instead of one year?

Why one week instead of one day?

ibrahimjaved12 replied:
Thanks @pdpinch

feoh commented:
Hello. The devops team is struggling a bit trying to get enough context around this change to review it effectively.

We're also concerned about the details in the testing section. We (as a wider team) should have clear instructions for testing this code given its importance.

I'll be following up with other OCW team members to see if I can get some of these questions answered. I realize this is a low content comment but I wanted you to know that we are looking and working on this.

feoh commented:
@ibrahimjaved12 - In order to facilitate thorough testing given this PR's complexity we will apply it to OCW RC this afternoon.

feoh commented:
Hi! Testing failed. Looks like you have a syntax error. See below:
```
# cpatti @ rocinante in ~/alt/ol-infrastructure/src/ol_infrastructure/applications/ocw_site on git:add-cache-control-headers-ocw o [16:40:11]
$ pr pulumi up -s applications.ocw_site.QA
Previewing update (applications.ocw_site.QA):
Downloading plugin: 16.58 MiB / 16.58 MiB [=========================] 100.00% 2s
                                                                                [resource plugin tls-4.11.1] installing
Downloading plugin: 18.15 MiB / 18.15 MiB [=========================] 100.00% 2s
                                                                                [resource plugin fastly-8.4.1] installing
Downloading plugin: 22.66 MiB / 22.66 MiB [=========================] 100.00% 3s
                                                                                [resource plugin consul-3.10.0] installing
Downloading plugin: 23.20 MiB / 23.20 MiB [=========================] 100.00% 3s
                                                                                [resource plugin github-5.22.0] installing
Downloading plugin: 23.68 MiB / 23.68 MiB [=========================] 100.00% 3s
                                                                                [resource plugin vault-5.17.0] installing
Downloading plugin: 27.27 MiB / 27.27 MiB [=========================] 100.00% 3s
                                                                                [resource plugin mongodbatlas-3.11.2] installing
Downloading plugin: 179.67 MiB / 179.67 MiB [=======================] 100.00% 5s
                                                                                [resource plugin aws-6.8.0] installing
     Type                        Name                                                             Plan       Info
     pulumi:pulumi:Stack         ol-infrastructure-ocw-site-application-applications.ocw_site.QA             2 warnings
 ~   ├─ pulumi:providers:fastly  fastly-provider                                                  update     [diff: ~version]
 ~   ├─ fastly:index:ServiceVcl  ocw-draft-qa                                                     update     [diff: ~snippets]
 ~   └─ fastly:index:ServiceVcl  ocw-live-qa                                                      update     [diff: ~snippets]

Diagnostics:
  pulumi:pulumi:Stack (ol-infrastructure-ocw-site-application-applications.ocw_site.QA):
    warning: geo_headers is deprecated: 'geo_headers' attribute has been deprecated and will be removed in the next major version release
    warning: geo_headers is deprecated: 'geo_headers' attribute has been deprecated and will be removed in the next major version release

Resources:
    ~ 3 to update
    13 unchanged

Do you want to perform this update? yes
Updating (applications.ocw_site.QA):
     Type                        Name                                                             Status                  Info
     pulumi:pulumi:Stack         ol-infrastructure-ocw-site-application-applications.ocw_site.QA  **failed**              1 error; 2 warnings
 ~   ├─ pulumi:providers:fastly  fastly-provider                                                  updated (4s)            [diff: ~version]
 ~   ├─ fastly:index:ServiceVcl  ocw-draft-qa                                                     **updating failed**     [diff: ~snippets]; 1 error
 ~   └─ fastly:index:ServiceVcl  ocw-live-qa                                                      **updating failed**     [diff: ~snippets]; 1 error

Diagnostics:
  fastly:index:ServiceVcl (ocw-draft-qa):
    error: 1 error occurred:
        * updating urn:pulumi:applications.ocw_site.QA::ol-infrastructure-ocw-site-application::fastly:index/serviceVcl:ServiceVcl::ocw-draft-qa: 1 error occurred:
        * invalid configuration for Fastly Service (6FCHJLMjrO50D4411UZrwA): Syntax error: Expected `{`, got `||`
    at: (input Line 858 Pos 105)
    if (bereq.url.path ~ ".*(main|common|www|course_v2|instructor_insights|fields)\.[0-9a-f]+\.(css|js)\z") || (bereq.url.path ~ ".*[0-9a-f]+\.[0-9a-f]+\.(css|js)\z") {
    --------------------------------------------------------------------------------------------------------##----------------------------------------------------------

  fastly:index:ServiceVcl (ocw-live-qa):
    error: 1 error occurred:
        * updating urn:pulumi:applications.ocw_site.QA::ol-infrastructure-ocw-site-application::fastly:index/serviceVcl:ServiceVcl::ocw-live-qa: 1 error occurred:
        * invalid configuration for Fastly Service (29GD8UCsPjRlNvgVp1SX2t): Syntax error: Expected `{`, got `||`
    at: (input Line 858 Pos 105)
    if (bereq.url.path ~ ".*(main|common|www|course_v2|instructor_insights|fields)\.[0-9a-f]+\.(css|js)\z") || (bereq.url.path ~ ".*[0-9a-f]+\.[0-9a-f]+\.(css|js)\z") {
    --------------------------------------------------------------------------------------------------------##----------------------------------------------------------

  pulumi:pulumi:Stack (ol-infrastructure-ocw-site-application-applications.ocw_site.QA):
    warning: geo_headers is deprecated: 'geo_headers' attribute has been deprecated and will be removed in the next major version release
    warning: geo_headers is deprecated: 'geo_headers' attribute has been deprecated and will be removed in the next major version release
    error: update failed

Outputs:
  - ocw_site_buckets: {
      - buckets                : [
      -     [0]: "ocw-content-draft-qa"
      -     [1]: "ocw-content-backup-draft-qa"
      -     [2]: "ocw-content-offline-draft-qa"
      -     [3]: "ocw-content-live-qa"
      -     [4]: "ocw-content-backup-live-qa"
      -     [5]: "ocw-content-offline-live-qa"
        ]
      - fastly_draft_service_id: "6FCHJLMjrO50D4411UZrwA"
      - fastly_live_service_id : "29GD8UCsPjRlNvgVp1SX2t"
      - policy                 : "ocw-site-content-read-only-qa20210504124911576700000001"
    }

Resources:
    ~ 1 updated
    11 unchanged

Duration: 30s

```

ibrahimjaved12 commented:
Thanks @feoh 
It looks that error came from the syntax of `if` condition combined with logical operator `||`.
I have fixed the syntax now.

Further I tested using `varnishd -C -f ./ttl_setup.vcl`, it was good for reporting syntax errors related to parentheses but that's all it did because it was expecting a different kind of file structure.

Also, using RE tools for `RE in Fastly` isn't entirely helpful because Fastly has minor syntax differences, as their docs say here:
> HINT: You can use the freely available [Regex101](https://regex101.com/) tool to test out your patterns, but be aware of potential minor syntax differences with Fastly's engine.
https://developer.fastly.com/reference/vcl/regex/

Even their given example does not work:` ^/admin(/.*)?\z ` because RE tools would expect something like `\/` instead of `/` and maybe `$` instead of `\z` .

However, I used this RE tester site: https://www.regextester.com/
Even though it reported syntax error because of forward slashes, but it still matched the strings, exactly in the way as they were intended to be. (Without `\z`) So I think REGEX wise we are all good.

I have fixed the IF conditions too. 
Let's try the testing again.
